README.md
# libdatetimetools
A library of tools for performing date and time calculations.

# Description
Description: libdatetimetools is a library of useful functions to enable
users to perform date and time calculations. One of its primary uses is as a
component of solutions to computing chain-based scheduling, as in litigation
and project management. 

# Version info
Version: 1.0
Created: 12/19/2020 12:43:04
Last Modified: Wed Dec 23 16:01:48 2020


# Usage:
Static library.

# File Formats:

Presently the library usese CSV files to import various rules
such as holidays.

# Restrictions:
Presently, the library handles Unix encoded ASCII files fine,
but chokes on macOS and Windows files. Handling CR and CRLFs will be
addressed in the future.

# Error Handling:
Under development

# License & Disclaimer
SPDX-License-Identifier:  LGPL-3.0-only

Author: Thomas H. Vidal (THV), thomashvidal@gmail.com
Organization: Dark Matter Computing

Copyright: (c) 2011-2020 - Thomas H. Vidal, Los Angeles, CA
All rights reserved

libdatetimetools is free software: you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by the
Free Software Foundation, version 3 of the License.
 
libdatetimetools is distributed in the hope that it will be useful,but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU General Public License along with
libdatetimetools.  If not, see <https://www.gnu.org/licenses/>.


CHANGELOG.md
## 1.0.0 (February 05, 2021)
  - Merge branch 'errorhandling' to bring fixes into master
  - Change date_difference parameters to structs; Add calendar-date unit test
  - Fixed file handling to deal with CRLF line endings
  - Added test for computing date using court days
  - Debugged courtday_difference; added testsuite
  - Initiate Error Handling Module
  
  ## 0.0.11 (January 11, 2021)
  - Added testsuite_interactive
  - Removed test_loop in favor of testsuite_run_check
  - Minor cleanups re MAXMESSAGELEN and commenting date_to_string
  - Added display manager to test suite; Added date_to_string
  - Improved data hiding; Improved function naming
  - Added the Xcode project
  - Add .gitattributes; modified .gitignore
  
  ## 0.0.10 (December 28, 2020)
  - Validating day_of_week precondition
  - Add new set_weekday function
  - Built initial unit test and test scripts
  - Update Makefile with install; start preparing timetools
  - Cleanups after the ALLMONTHS merge.
  
  ## 0.0.9 (December 26, 2020)
  - Merge pull request #1 from Dark-Matter-Computing/ALLMONTHS
  - Merge branch 'master' into ALLMONTHS
  - Refactored with ALLMONTHS at 0 and tested. Results were clear.
  - Refactored the code to define ALLMONTHS as 0 instead of 13.
  - neglected to add the file.
  - created version.h
  
  ## 0.0.8 (December 23, 2020)
  - Cleaned up the internal header file. Renamed the weekday fuction to something that made more sense for users. Added range checking for the weekday fuction. Began cleaning up comments.
  - Removed the diagnostic printfs after debugging the checkrule function.
  - Debugged the checkrule function.
  - removed some remaining diagnostic printif's remaining from debugging the holiday hash table.
  - Removed diagnostic printf commands after debugging the holiday hash table.
  - Finished debugging the holiday hash table, before unit testing.
  - Debugging the add_rule function that builds that Holiday hash table.
  - Debugged the processtoken function
  - Debugged the get_fields function
  - Removed diagnostic printf commands.
  - Debugged holiday_rules_get_tokens function.
  
  ## 0.0.7 (December 22, 2020)
  - Addressed all compilation errors after adding holiday rule processing. Debugged the tokenizer.
  - Refactored the code to bring the holiday rule lexical analysis into the library. Added a private header and moved some data types and functions into the private header.
  
  ## 0.0.6 (December 18, 2020)
  - Refactored by bringing certain functions from DocketMaster's former ruleprocessor module into the library (where they belong). Updated the README.
  
  ## 0.0.5 (December 15, 2020)
  - renamed the struct DATETIME to DateTime to comply with naming conventions for structs.
  
  ## 0.0.4 (December 10, 2020)
  - Began refactoring code, beginning with the comments.
  
  ## 0.0.3 (November 28, 2020)
  - Rename COPYING.LESSER to LICENSE for clarity.
  - Automated makefile. Separated timetools prototypes into timetools.h.
  - Makefile + Cleanup + Fixme
  
  ## 0.0.2 (November 25, 2020)
  - Pre-release vers - Changed organization name to Dark Matter Computing
  - Pre-release vers - Added LGPL License
  - Pre-release vers - 2020-Jul-03
  
  ## 0.0.1 (October 22, 2011)
  - Pre-release vers - added timetools - 2011-Oct-22
  - Pre-release vers - added timetools - 2011-Oct-20
  - Pre-release vers - added timetools - 2011-Oct-09
  - Pre-release vers - see ChangesLog - 2011-Oct-09
  - Pre-release vers - see ChangesLog - 2011-Oct-09
  - Pre-release vers - see ChangesLog - 2011-Oct-09
  - Pre-release vers - see ChangesLog - 2011-Oct-09
  - Pre-release vers - see ChangesLog - 2011-Oct-09
  - Pre-release vers - see ChangesLog - 2011-Oct-09
  - Pre-release vers - see ChangesLog - 2011-Oct-09
  - Pre-release vers -  2011-Oct-09
  - Pre-release vers - islastweek
  
  ## 0.0.0 (August 18, 2011)
  - initial project version - created on 2011-Aug-18




src/datetimetools_pvt.h
/*
 * Filename: datetimetools_int.h
 * Library: libdatetimetools
 *
 * Description: libdatetimetools is a library of useful functions to enable
 * users to perform date and time calculations. One of its primary uses is as a
 * component of solutions to computing chain-based scheduling, as in litigation
 * and project management. 
 *
 * Version: See VERSION
 * Created: 12/19/2020 12:43:04
 * Last Modified: Fri Feb  5 18:07:12 2021
 *
 * Author: Thomas H. Vidal (THV), thomashvidal@gmail.com
 * Organization: Dark Matter Computing
 *
 * Copyright: (c) 2011-2020 - Thomas H. Vidal, Los Angeles, CA
 * SPDX-License-Identifier:  LGPL-3.0-only
 *
 * Usage: 
 * File Format: Presently the library usese CSV files to import various rules
 * such as holidays.
 * 
 * Restrictions: Presently, the library handles Unix encoded ASCII files fine,
 * but chokes on macOS and Windows files. Handling CR and CRLFs will be
 * addressed in the future.
 * 
 * Error Handling: Under development
 * 
 * Notes: this private header handles all the data type and file management
 * activities, which should be unnecessary to users of the library.
 */

#ifndef _DATETIMETOOLS_PVT_H_INCLUDED_
#define _DATETIMETOOLS_PVT_H_INCLUDED_

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <math.h>
#include "../include/datetools.h"

/*-----------------------------------------------------------------------------
 * PARAMETERIZED MACROS
 *----------------------------------------------------------------------------*/

#define SET_FLAG(x,y) ((x) |= (y))
    /* SET_BIT sets the bit in x corresponding to the event flag passed by
    the y parameter. */

#define CLEAR_FLAG(x,y) ((x) &= ~(y))
    /* CLEAR_BIT takes x and clears from x ALL the bits that are set in y. */

#define TEST_FLAG(x,y) ((x) & (y))
    /* TEST_BIT tests a single bit to see if it's "on." This will not test to
    see if MULTIPLE bits are set. */

#define TEST_MULTIPLEFLAGS(x,y) ((x) & ~(y))
    /* TEST_MULTIPLEFLAGS tests whether bits IN ADDITION TO the relevant bit
    are set */

#define CLEAR_ALLFLAGS(x) ((x) &= 0)
    /* CLEARS_ALLFLAGS zeros out the flag variable. */

#define ASCII2DECIMAL(x) ((x)-48)

    /* ASCII2DECIMAL: convert an ASCII character into a decimal. Macro does no
     * error checking (of course!), but anticipates an ascii character between
     * '0' ASCII 48) and 9 (ASCII 57).
     */


/*-----------------------------------------------------------------------------
 * Symbolic Constants: Holiday File Field Codes 
 *----------------------------------------------------------------------------*/

/*  Define field codes for the holiday CSV File */

const char *HF_MONTH     = "Month";
const char *HF_RTYPE     = "Rule Type";
const char *HF_RULE      = "Rule";
const char *HF_HOLIDAY   = "Holiday";
const char *HF_AUTHORITY = "Authority";

/*  Sizes and numbers of records and fields */

#define MAXRECORDLENGTH 500 /* Maximum Length (in characters of CSV File line,
                               each line consists of one record. */
#define MAXNUMFIELDS 25 /* Maxinum number of fields in CSV File */
#define MAXFIELDLEN 25 /* Maximum length (in chars) of name of field */

/*  Field and Text delimiters */

#define FDELIMITER ',' /* Field delimiter used in the CSV files */
#define TDELIMITER '\"' /* Text string delimiter used in the CSV files */
#define NEWLINE '\n' /*  Newline */
#define CARRIAGE_RTN '\r' /* carriage return */
#define NULCHAR '\0' /* nul string */
#define EMPTYFIELD '\0'

/*  Token processing flags */

const unsigned char BEGIN_FIELD = (1<<0); /* Set when the first field delimiter is reached. */
const unsigned char BEGIN_TSTRING = (1<<1); /* Set when the first text delimiter is reached. */
const unsigned char END_TSTRING = (1<<2); /* Set when the end text delimter is reached. */
const unsigned char TOKEN_FOUND = (1<<7); /* Set when the token was found */

/* Macro definitions for error codes */

#define NOERROR 0
#define BADCOMMANDLINE 1
#define NULSTRING 2
#define NOFDELIM 3
#define NOTDELIM 4

/*-----------------------------------------------------------------------------
 * EXPORTED, BUT PRIVATE, DATA TYPES 
 *----------------------------------------------------------------------------*/

/* rulefile state */
struct RuleSet {
    FILE *rulefile;
    char headerfields[MAXNUMFIELDS][MAXFIELDLEN];
    int totalnumfields;
    enum {CLOSED, OPEN} openstatus;
};

/* The holiday rules are maintained in a chained hash table data structure.
 * The hash table is an array of linked lists. The array of nodes is indexed to
 * the months of the year.  Each pointer is the first node of a linked list of
 * HolidayNodes.  Each element holds the rules for the applicable month.  There
 * are 13 elements as the final element applies to the linked lists of rules
 * that apply to every month, like weekend rules.  The hash function is thus
 * the months of the year. 
 */

struct HolidayRule{
    int month;
    char ruletype;
        /* holds one of three possible values: 'w' for a weekend rule, 'r'
         * for a relative rule, and 'a' for an absolute rule
         */
    unsigned int wkday;
        /* the day of the wk to which the rule applies. if the rule is
         * relative.
         */
    int wknum;
        /* the week number to which the rule applies if the rule is
         * relative.
         */
    int day; /* the day of the holiday if the ruletype is 'a' */
    char holidayname[40]; /* the title of the holiday */
    char authority[100]; /* the statutory authority for the holiday */
};

struct HolidayNode { /* structure for linked list to hold court holidays */
    struct HolidayRule rule; 
    struct HolidayNode *nextrule; 
};

/* The hash table itself: an array to hold an individual holiday-node linked
list for each month of the year + the ALLMONTHS rules */
extern struct HolidayNode *holidayhashtable[13];

/*-----------------------------------------------------------------------------
 * Holiday Hashtable Handler Functions
 *----------------------------------------------------------------------------*/

void holiday_tbl_build(FILE *receivedrulefile,
                       struct HolidayNode *holidayhashtable[]);
void holiday_tbl_init(struct HolidayNode *holidayhashtable[]);
void holiday_rules_get_tokens(FILE *holidayrulefile,
                              struct HolidayNode *holidayhashtable[],
                              struct RuleSet *globalstate);
char *holiday_rules_tokenize(char *string, int *lasttoken);
void holiday_rules_parse_token(char *token, char *cur_field,
                                struct HolidayRule *newholiday);
void holiday_table_addrule(struct HolidayNode **elementhandle,
                                          struct HolidayRule *newrule);
void holiday_table_release(struct HolidayNode *holidayhashtable[]);

/*-----------------------------------------------------------------------------
 * Holiday Rule File Management 
 *----------------------------------------------------------------------------*/

int holiday_rules_validatefile(FILE *candidaterulefile_h);
void holiday_rules_getfields(FILE *holidayrulefile, struct RuleSet *globalstate);
void holiday_rules_resetfile(FILE *holidayrulefile);
int holiday_rules_closefile(FILE *holidayrulefile);

/*-----------------------------------------------------------------------------
 * Process Holiday Rules
 *----------------------------------------------------------------------------*/

int holiday_tbl_checkrule(struct DateTime *dt, struct HolidayNode *rulenode);

/*-----------------------------------------------------------------------------
 *  Error Handling
 *----------------------------------------------------------------------------*/

void errorprocessor(int error_code);

#endif /* _DATETIMETOOLS_PVT_H_INCLUDED_ */
